solver/src/solver/http_solver/model.rs

```diff
@@ -26,44 +28,54 @@ pub struct OrderModel {
 }
 
 #[derive(Debug, Serialize)]
-pub struct UniswapModel {
-    pub token1: String,
-    pub token2: String,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    pub balance1: u128,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    pub balance2: u128,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    pub fee: f64,
+pub struct PoolTokenData {
+    #[serde(with = "u256_decimal")]
+    pub balance: U256,
+    #[serde(with = "u256_decimal")]
+    pub weight: U256,
+}
+
+#[derive(Debug, Serialize)]
+pub enum PoolType {
+    UniswapV2,
+}
```

Comment:
I think we can use a tagged enum so that we can model different pools that have completely different parameters. While technically a Uniswap pool is a weighted pool with weights of `1.0` - we might not necessarily be able to find a model that is compatible with all of Uniswap pools, Balancer wieghted pools, and other kinds of AMMs (like Balancer stable pools for example).

Something along the lines of:
```rust
pub struct PoolModel {
    pub parameters: PoolParameters,
    pub cost: CostModel,
    pub mandatory: bool,
}

pub enum PoolParameters {
    UniswapV2(UniswapPoolParameters),
    BalancerWeighted(BalancerWeightedPoolParameters),
}

pub struct UniswapPoolParameters {
    pub reserves: HashMap<H160, U256>,
    pub fee: BigDecimal,
}

pub struct BalancerWeightedPoolParameters {
    pub reserves: HashMap<H160, BalancerWeightedPoolTokenData>,
    pub fee: BigDecimal,
}
```

Reply:
This all seems like a good idea, but we need to keep the solver team in the loop on this so they know what to expect. @marcovc and @twalth3r

Reply:
This was the changes we discussed though right? That we would used a "tagged union" in order to represent pools so we can have different parameters/structure for different kinds of pools that don't fit into the weighted pool model.

Reply:
This will output JSON looking like:
```json
[
  {
    "kind": "UniswapV2",
    "reserves": {
      "0x0102030405060708091011121314151617181920": "100000"
    },
    "fee": "0.003",
    "cost": {
      "amount": "1337",
      "token": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
    },
    "mandatory": false
  },
  {
    "kind": "BalancerWeighted",
    "reserves": {
      "0x0102030405060708091011121314151617181920": {
        "balance": "100000",
        "weight": "0.5"
      }
    },
    "fee": "0.003",
    "cost": {
      "amount": "1337",
      "token": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
    },
    "mandatory": false
  }
]
```

Reply:
I got the impression that you were suggesting to introduce a "parameters" field with contained the pools inner information.

Reply:
I now see the "flatten"